### PR TITLE
Restore 100% Python coverage

### PR DESCRIPTION
**Pull request by dsfaccini**
> This pull request was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Follow-up to #536 and [the review question about its 52% pipeline threshold](https://github.com/pydantic/genai-prices/pull/536#discussion_r3842240951).

## Summary

- Bring `prices/src/prices/**` and `tests/dataset/**` to 100% statement and branch coverage.
- Replace the package/pipeline coverage split with one unfiltered 100% Python gate in coverage.py, Make, and CI.
- Exercise importer orchestration, build/export validation, maintenance commands, discrepancy handling, provider types, and the dataset regeneration paths with isolated tests.
- Make the AWS source updater import-safe by creating live clients and fetching models only inside `main()`.
- Update `AGENTS.md` to describe the unified gate and the existing JavaScript Make target accurately.

## Verification

- Python tests: pass; required all-passing and 100%-coverage bars met.
- Ruff and basedpyright strict: pass.
- JavaScript CI: pass.
- Full pre-commit hooks: pass for every PR file.
- `make build`: pass with no generated artifact changes.
- Independent review: approved after strengthening the AWS script-entrypoint assertion.

## Scope decisions

- The JavaScript coverage thresholds are unchanged. This PR removes the Python pipeline exception questioned on #536.
- The AWS updater is retained and made testable rather than deleted as unused code.
- Published pricing data and generated package data are unchanged.

<a href="https://cubic.dev/pr/pydantic/genai-prices/pull/598?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -98,8 +96,7 @@ skip_covered = true
 show_missing = true
 ignore_errors = true
 precision = 2
-# Bars live on the filtered reports (package 100 / pipeline 52), not on an unfiltered total.
-fail_under = 0
+fail_under = 100
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply pyproject.toml:99]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Dismissed: the exact unfiltered coverage job passed at 100% on the reviewed head. `tests/dataset/get_raw_bodies.py` was not reported as an uncovered measured module, so the predicted gate failure did not occur.

File: tests/test_dataset.py [hunk 2]
```diff
@@ -36,3 +42,211 @@ def test_usages_dataset_is_up_to_date(extract_usages_module: object) -> None:
...
+    current: list[dict[str, object]] = []
+    rebuilt: list[dict[str, object]] = [{'body': {'file': 'recorded.json'}}]
+
+    def stale_result() -> tuple[list[dict[str, object]], list[dict[str, object]]]:
+        return current, rebuilt
+
+    monkeypatch.setattr(extract_usages_module, 'rebuild_usages', stale_result)
+
+    with pytest.raises(AssertionError, match='usages.json is out of date'):
+        extract_usages_module.main(write=False)  # pyright: ignore[reportAttributeAccessIssue, reportUnknownMemberType]
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply tests/test_dataset.py:61]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Accepted and fixed in b927a00. The no-write case now uses a temporary dataset file and asserts its sentinel contents remain unchanged after the error.

File: tests/test_dataset.py [hunk 3]
```diff
@@ -36,3 +42,211 @@ def test_usages_dataset_is_up_to_date(extract_usages_module: object) -> None:
...
+
+    monkeypatch.setattr(extract_usages_module, 'this_dir', tmp_path)
+    monkeypatch.setattr(extract_usages_module, 'raw_bodies_path', raw_bodies)
+    monkeypatch.setattr(extract_usages_module, 'get_usages', fake_get_usages)
+
+    rebuilt_current, rebuilt = extract_usages_module.rebuild_usages()  # pyright: ignore[reportAttributeAccessIssue, reportUnknownMemberType, reportUnknownVariableType]
+
+    assert rebuilt_current == current
+    expected_bodies = recordings if has_raw_bodies else [{'file': 'current.json'}]
+    assert calls == [expected_bodies, expected_bodies] if has_raw_bodies else [expected_bodies]
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply tests/test_dataset.py]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Accepted and fixed in b927a00 by parenthesizing the conditional expected value, so the fallback branch now verifies the recorded calls.

File: tests/test_price_discrepancies.py [hunk 4]
```diff
@@ -22,3 +84,248 @@ def test_prices_conflict_does_not_raise_on_extra_source_key() -> None:
...
+    second_affected = model('also-affected', ModelPrice(input_mtok=Decimal('1')))
+    second_affected.price_discrepancies = {'source': {}}
+    provider_ymls = {
+        'first': FakeProviderYaml(provider('first', [affected, model('clear', ModelPrice(input_mtok=Decimal('1')))])),
+        'second': FakeProviderYaml(provider('second', [second_affected])),
+        'third': FakeProviderYaml(provider('third', [model('clear', ModelPrice(input_mtok=Decimal('1')))])),
+    }
+    monkeypatch.setattr(price_discrepancies, 'get_providers_yaml', loaded_providers(provider_ymls))
+
+    assert check_for_price_discrepancies() == 2
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply tests/test_price_discrepancies.py]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Accepted and fixed in b927a00. The test now puts two affected models under one provider, and `check_for_price_discrepancies` returns the documented discrepancy total rather than the number of affected providers.

File: prices/src/prices/price_discrepancies.py [hunk 5]
```diff
@@ -142,7 +152,7 @@ def check_for_price_discrepancies() -> int:
         if discs:
             if not found:
                 print('price discrepancies:')
-            found += 1
+            found += discs
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply prices/src/prices/price_discrepancies.py:155]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Accepted and fixed in 18c7480. `check_for_price_discrepancies` now returns a bounded 0/1 process status while the printed report retains the actual per-provider discrepancy counts.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -91,17 +91,14 @@ jobs:
           enable-cache: true # zizmor: ignore[cache-poisoning] -- Job does not produce release artifacts and does not have sensitive permissions
       - run: uv sync --only-dev
       - run: uv run coverage combine coverage
-      # Smokeshow / the public badge are the published package at 100%. Pipeline ratchet is the next step.
-      - run: uv run coverage html --include='packages/python/*,tests/*' --omit='tests/dataset/*' --show-contexts --title "genai-prices coverage for ${{ github.sha }}"
+      - run: uv run coverage html --show-contexts --title "genai-prices coverage for ${{ github.sha }}"
```
Comment (Kludex) [inline .github/workflows/ci.yml:94]: We can remove the html thing in a follow-up

File: prices/src/prices/price_discrepancies.py [hunk 7]
```diff
@@ -132,22 +142,22 @@ def check_for_price_discrepancies() -> int:
...
             if not found:
                 print('price discrepancies:')
-            found += 1
+            found = True
             print(f'{provider_yml.provider.name:>20}: {discs}')
 
     if not found:
         print('no price discrepancies found')
-    return found
+    return int(found)
```
Comment (Kludex) [inline prices/src/prices/price_discrepancies.py]: I'm pretty sure the reason was to count... now it can only be 1 or 0

File: .github/workflows/ci.yml (hunk 6 above)
Comment (dsfaccini) [reply .github/workflows/ci.yml:94]: wdym? it's not removing the HTML thing, just removing the `--include ...` option

or did you mean "let's remove the HTML thing in a follow-up"?

Comment (Kludex) [reply .github/workflows/ci.yml:94]: I mean that coverage html smokeshow is useless when you have 100% coverage, we can remove that line

File: prices/src/prices/price_discrepancies.py (hunk 7 above)
Comment (dsfaccini) [reply prices/src/prices/price_discrepancies.py]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

You are right. The count was the function contract; normalizing it inside the function solved the exit-status issue at the wrong layer. Fixed in dbe90da: the function returns the actual discrepancy count, while only the CLI command maps zero to exit 0 and any positive count to exit 1. Tests cover counts 0, 3, and 256.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (dsfaccini) [reply .github/workflows/ci.yml:94]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

Agreed — keeping removal of the published HTML report as a separate follow-up.

**Comment by Kludex** [comment]
Can you tell your claude to avoid testing low level interfaces? Focus on high level testing, and to try to reach internal code from that. If the test is too complex to create because of edge cases, then a pragma will do.

**Comment by Kludex** [comment]
Reworked the coverage additions to exercise public commands and importer entry points through real temporary files. Low-level helper suites were removed, and narrowly justified pragmas now cover only unreachable or fixture-heavy branches; `make test` remains at 100% coverage.

File: tests/test_maintenance_commands.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+"""
+    )
+    provider_yaml = ProviderYamlFile(provider_path)
+
+    monkeypatch.setattr(collapse, 'get_providers_yaml', lambda: {'example': provider_yaml})
+    collapse.collapse()
+    assert capsys.readouterr().out == 'Provider example:\n  1 models combined\n\nTotal models combined: 1\n'
+    saved = ProviderYamlFile(provider_path)
+    assert [model.id for model in saved.provider.models] == ['not-collapsible', 'parent', 'parent:different']
+    assert saved.provider.find_model('parent:matching') is not None
```
[the remark this answers is not in the document]
Comment (Kludex) [reply tests/test_maintenance_commands.py:74]: Dismissed: the child YAML row is removed, but `collapse()` first merges its matcher into the parent. The positive lookup passes and verifies that `parent:matching` remains a valid alias of `parent` after the collapse.

File: prices/src/prices/update.py [hunk 9]
```diff
@@ -104,7 +104,9 @@ def add_id_to_model(self, lookup_id: str, new_model_id: str) -> None:
         model = model.model_copy(update=dict(match=ClauseEquals(equals=new_model_id)))
         self.update_model(lookup_id, model)
 
-    def set_price_discrepency(self, lookup_id: str, source: str, price: ModelPrice) -> None:
+    def set_price_discrepency(  # pragma: no cover - discrepancy writes are deliberately disabled in the command
```
[the remark this answers is not in the document]
Comment (Kludex) [reply prices/src/prices/update.py]: Accepted and fixed in a5dece2. The discrepancy-writing workflow is disabled, so the unused helper was removed instead of excluded from coverage.

File: tests/test_maintenance_commands.py
Comment (Kludex) [inline tests/test_maintenance_commands.py:1]: I don't even know what this is for

Comment (dsfaccini) [reply tests/test_maintenance_commands.py:1]: the file?

image

**Comment by dsfaccini** [comment]
> Can you tell your claude to avoid testing low level interfaces? Focus on high level testing, and to try to reach internal code from that. If the test is too complex to create because of edge cases, then a pragma will do.

I can open a PR to enhance/refine the instructions in the repo

File: tests/test_maintenance_commands.py
Comment (Kludex) [reply tests/test_maintenance_commands.py:1]: yes

Comment (dsfaccini) [reply tests/test_maintenance_commands.py:1]: > This reply was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

It covers the internal price-maintenance commands (`collapse`, deprecation detection, provider-list injection) and the `python -m prices` dispatcher. The name was too vague, so I renamed it to `test_maintenance_commands.py` in 7d23a37.
